File: src/main/java/fi/quanfoxes/Phase.java
package fi.quanfoxes;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

/**
 * Represents one phase in compilation
 */
public abstract class Phase {
    private ExecutorService executors;
    private boolean enabled;

    private List<Future<Status>> tasks = new ArrayList<>();

    /**
     * Represents a completed task
     * @param <T> Type of value the task produced
     */
    private class CompletedTask<T> implements Future<T> {
        private T value;

        /**
         * Represents a task that produced the given value
         * @param value Value that the task produced
         */
        public CompletedTask(T value) {
            this.value = value;
        }

        /**
         * Cancels the task
         */
        public boolean cancel(boolean b) {
            return false;
        }

        /**
         * Returns the value that the task produced
         */
        public T get() throws InterruptedException, ExecutionException {
            return value;
        }

        /**
         * Returns the value that the task produced
         */
        public T get(long time, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
            return value;
        }

        /**
         * Returns whether the task was canceled
         */
        public boolean isCancelled() {
            return false;
        }

        /**
         * Returns whether the task has finished
         */
        public boolean isDone() {
            return true;
        }
    }

    /**
     * Sets the executors for this phase
     * 
     * @param executors Executors to use during this phase
     * @param enabled   Should this phase use multi-threading
     */
    public void setExecutors(ExecutorService executors, boolean enabled) {
        this.executors = executors;
        this.enabled = enabled;
    }

    /**
     * Executes the phase with the given data
     * 
     * @param bundle Data collection that the phase may need
     */
    public abstract Status execute(Bundle bundle);

    /**
     * Executes runnable on another thread if multithreading is enabled, otherwise
     * executes locally
     * 
     * @param runnable Runnable to run
     */
    public void async(Callable<Status> runnable) {
        if (enabled) {
            tasks.add(executors.submit(runnable));
        } else {
            Status status = null;

            try {
                status = runnable.call();
            } catch (Exception e) {
                status = Status.error(e.getMessage());
            }

            tasks.add(new CompletedTask<Status>(status));
        }
    }

    /**
     * Waits for all tasks to finish
     */
    public void sync() {
        int i = 0;

        while (i < tasks.size()) {
            Future<Status> task = tasks.get(i++);
            while (!task.isDone());
        }
    }

    /**
     * Returns whether all tasks succeeded during the phase
     * @return True if all tasks succeeded during the phase
     */
    public boolean succeeded() {
        return !tasks.stream().filter((t) -> {
            try {
                return !t.isDone() || t.get().isProblematic();
            }
            catch (Exception e) {
                return false;
            }
        }).findFirst().isPresent();
    }

    /**
     * Aborts the execution
     */
    public void abort() {
        System.exit(1);
    }
}

File: Zigzag/Phase.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Text;

/// <summary>
/// Represents one phase in compilation
/// </summary>
public abstract class Phase
{
	private List<Task<Status>> Tasks { get; } = new List<Task<Status>>();
	public bool Multithread { get; set; }

	public bool Failed => Tasks.Any(t => !t.IsCompleted || t.Result.IsProblematic);

	/// <summary>
	/// Executes the phase with the given data 
	/// </summary>
	/// <param name="bundle">Data collection that the phase may need</param>
	/// <returns>Status returned from the phase</returns>
	public abstract Status Execute(Bundle bundle);

	/// <summary>
	/// Executes runnable on another thread if multithreading is enabled, otherwise executes locally
	/// </summary>
	/// <param name="task">Task to run</param>
	public void Run(Func<Status> task)
	{
		if (Multithread)
		{
			Tasks.Add(Task.Run(task));
		}
		else
		{
			Status status;

			try
			{
				status = task();
			}
			catch (Exception e)
			{
				status = Status.Error(e.Message);
			}

			Tasks.Add(Task.FromResult(status));
		}
	}

	/// <summary>
	/// Waits for all tasks to finish
	/// </summary>
	public void Sync()
	{
		var i = 0;

		while (i < Tasks.Count)
		{
			Tasks[i++].Wait();
		}
	}

	/// <summary>
	/// Returns all tasks errors that occured during execution
	/// </summary>
	public string GetTaskErrors()
	{
		var builder = new StringBuilder("\n");

		foreach (var task in Tasks)
		{
			if (task.Result.IsProblematic)
			{
				builder.Append(task.Result.Description).Append('\n');
			}
		}

		return builder.ToString();
	}


	/// <summary>
	/// Aborts the execution
	/// </summary>
	public static void Abort()
	{
		Environment.Exit(1);
	}
}